File: Baekjoon/유석환/Q11053.swift
let N = Int(readLine()!)!
var A = readLine()!.split(separator: " ").map { Int(String($0))! }
var dp = [Int](repeating: 0, count: N)
dp[0] = 1

for i in 1..<N {
    var maxDp = 0 // 자신보다 작은 값 중 최대 dp값
    for j in 0..<i {
        if A[i] > A[j] {
            maxDp = max(dp[j], maxDp)
        }
    }
    dp[i] = maxDp + 1
}

print(dp.max()!)

File: Baekjoon/노수일/Q11053.py
from sys import stdin

n = int(stdin.readline())
numbers = [int(num) for num in stdin.readline().split()]
memo = [0 for _ in range(n)]

for i in range(n):
    for j in range(i):
        if numbers[i] > numbers[j] and memo[j] > memo[i]:
            memo[i] = memo[j]
    memo[i] += 1

print(memo)
print(max(memo))

"""
8
50 60 30 10 30 20 40 50
 0  0  0  0  0  0  0  0
 1  0  0  0  0  0  0  0
 1  2  0  0  0  0  0  0
 1  2  1  0  0  0  0  0
 1  2  1  1  0  0  0  0
 1  2  1  1  2  0  0  0
 1  2  1  1  2  1  0  0
 1  2  1  1  2  2  0  0 
"""
